pkg/conversion/capi2mapi/openstack.go

```diff
@@ -0,0 +1,701 @@
+func (m machineSetAndOpenStackMachineTemplateAndOpenStackCluster) ToMachineSet() (*mapiv1.MachineSet, []string, error) { //nolint:dupl
+	if m.machineSet == nil || m.template == nil || m.openstackCluster == nil || m.machineAndOpenStackMachineAndOpenStackCluster == nil {
+		return nil, nil, errCAPIMachineSetOpenStackMachineTemplateOpenStackClusterCannotBeNil
+	}
+
+	var (
+		errors   []error
+		warnings []string
+	)
+
+	// Run the full ToMachine conversion so that we can check for
+	// any Machine level conversion errors in the spec translation.
+	mapiMachine, warns, err := m.ToMachine()
+	if err != nil {
+		errors = append(errors, err)
+	}
+
+	warnings = append(warnings, warns...)
+
+	mapiMachineSet, err := fromCAPIMachineSetToMAPIMachineSet(m.machineSet)
+	if err != nil {
+		errors = append(errors, err)
+	}
+
+	if len(errors) > 0 {
+		return nil, warnings, utilerrors.NewAggregate(errors)
+	}
+
+	mapiMachineSet.Spec.Template.Spec = mapiMachine.Spec
+
+	// Copy the labels and annotations from the Machine to the template.
+	mapiMachineSet.Spec.Template.ObjectMeta.Annotations = mapiMachine.ObjectMeta.Annotations
+	mapiMachineSet.Spec.Template.ObjectMeta.Labels = mapiMachine.ObjectMeta.Labels
+
+	return mapiMachineSet, warnings, nil
+}
+
+// Conversion helpers.
+
+func convertCAPOAdditionalBlockDevicesToMAPO(fldPath *field.Path, capoAdditionalBlockDevices []capov1.AdditionalBlockDevice) ([]mapiv1alpha1.AdditionalBlockDevice, field.ErrorList) {
+	mapoAdditionalBlockDevices := []mapiv1alpha1.AdditionalBlockDevice{}
+	errors := field.ErrorList{}
+
+	for i, capoAdditionalBlockDevice := range capoAdditionalBlockDevices {
+		mapoAdditionalBlockDevice := mapiv1alpha1.AdditionalBlockDevice{
+			Name:    capoAdditionalBlockDevice.Name,
+			SizeGiB: capoAdditionalBlockDevice.SizeGiB,
+			Storage: mapiv1alpha1.BlockDeviceStorage{
+				Type: mapiv1alpha1.BlockDeviceType(capoAdditionalBlockDevice.Storage.Type),
+			},
+		}
+
+		if capoAdditionalBlockDevice.Storage.Type == capov1.VolumeBlockDevice {
+			if capoAdditionalBlockDevice.Storage.Volume == nil {
+				errors = append(errors, field.Invalid(fldPath.Index(i), capoAdditionalBlockDevice, "The volume field must be populated if type is volume"))
+			} else {
+				mapoAdditionalBlockDevice.Storage.Volume = &mapiv1alpha1.BlockDeviceVolume{
+					Type: capoAdditionalBlockDevice.Storage.Volume.Type,
+				}
+
+				if capoAdditionalBlockDevice.Storage.Volume.AvailabilityZone.From == capov1.VolumeAZFromName {
+					mapoAdditionalBlockDevice.Storage.Volume.AvailabilityZone = string(*capoAdditionalBlockDevice.Storage.Volume.AvailabilityZone.Name)
+				}
+			}
+		}
+
+		mapoAdditionalBlockDevices = append(mapoAdditionalBlockDevices, mapoAdditionalBlockDevice)
+	}
+
+	return mapoAdditionalBlockDevices, errors
+}
+
+func convertCAPOOpenStackIdentityReferenceToMAPO(capoIdentityRef *capov1.OpenStackIdentityReference) (string, *corev1.SecretReference) {
```

Comment:
Is there a symmetric function to go the other way? Or is that not needed with OpenStack handling?

Reply:
There is. It's `convertMAPOCloudNameSecretToCAPO` in `pkg/conversion/mapi2capi/openstack.go`.